### PR TITLE
POC Using dagger for Config

### PR DESCRIPTION
**Pull request by tylerbenson**
Includes a test that mocks the config interface.

### CODE REVIEW

File: internal-api/src/main/java/datadog/trace/bootstrap/config/provider/ConfigConverter.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  /**
+   * @param value to parse by tClass::valueOf
+   * @param tClass should contain static parsing method "T valueOf(String)"
+   * @param <T>
+   * @return value == null || value.trim().isEmpty() ? defaultValue : tClass.valueOf(value)
+   * @throws NumberFormatException
+   */
+  static <T> T valueOf(final String value, @NonNull final Class<T> tClass) {
+    if (value == null || value.trim().isEmpty()) {
```
Comment (richardstartin) [inline internal-api/src/main/java/datadog/trace/bootstrap/config/provider/ConfigConverter.java:20]: Consider using `ClassValue` here - in fact that's what it was intended for - reasonably fast dynamic augmentation of types.

File: internal-api/src/main/java/datadog/trace/bootstrap/config/provider/EnvironmentConfigProvider.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package datadog.trace.bootstrap.config.provider;
+
+public class EnvironmentConfigProvider implements ConfigProvider.Source {
```
Comment (dougqh) [inline internal-api/src/main/java/datadog/trace/bootstrap/config/provider/EnvironmentConfigProvider.java:3]: To me it is strange that these are called ConfigProviders when they are really Config Sources.

Comment (tylerbenson) [reply internal-api/src/main/java/datadog/trace/bootstrap/config/provider/EnvironmentConfigProvider.java:3]: Yeah, the name is easy to change.  It started as something else then evolved while I was working on it and forgot to rename the class.

File: internal-api/src/main/java/datadog/trace/bootstrap/config/provider/ConfigConverter.java [hunk 3]
```diff
@@ -21,18 +23,27 @@
...
-    } catch (final NoSuchMethodException | IllegalAccessException e) {
-      log.debug("Can't invoke or access 'valueOf': ", e);
-      throw new NumberFormatException(e.toString());
     } catch (final Throwable e) {
       log.debug("Can't parse: ", e);
       throw new NumberFormatException(e.toString());
     }
   }
+
+  private static class ValueOfLookup extends ClassValue<MethodHandle> {
```
Comment (richardstartin) [inline internal-api/src/main/java/datadog/trace/bootstrap/config/provider/ConfigConverter.java:35]: This is like a dynamic type, this is awesome, I wonder what else we could use things like this for.

**Comment by dougqh** [comment]
My rough idea of what a better Config API might look like is...

`
@Config(namespace="jetty")
interface JettyConfig {
  @Config.Int(name="threshold", min=0, max=100, default=0)
  int threshold();
}

class JettyDecorator {
  static final int THRESHOLD = Config.get(JettyConfig.class).threshold();
}`

That said, I actually think a PR isn't really the best way to move design forward.
I think constructing a straw man in advance would work better.

But I would still like to see the ConfigProvider-s split out incorporated into Config, that seems like a pretty clear win.

**Comment by tylerbenson** [comment]
Closing this POC PR for now.  We can reopen later if we want to revisit Dagger.

**Pull request closed without merging**